trio_run_in_process/run_in_process.py

```diff
@@ -193,17 +194,40 @@ async def _open(
                         # If a keyboard interrupt is encountered relay it to the
                         # child process and then give it a moment to cleanup before
                         # re-raising
+                        logger.debug(
+                            "Got KeyboardInterrupt, sending SIGINT to %s", proc
+                        )
                         try:
                             proc.send_signal(signal.SIGINT)
-                            with trio.move_on_after(2):
-                                await proc.wait()
+                            try:
+                                with trio.fail_after(constants.SIGINT_TIMEOUT_SECONDS):
+                                    await proc.wait()
+                            except trio.TooSlowError:
+                                logger.debug(
+                                    "Timed out waiting for %s to exit after relaying SIGINT",
+                                    proc,
+                                )
                         finally:
                             raise err
+                    else:
+                        await proc.wait()
                 finally:
-                    await proc.wait()
-                    logger.debug(
-                        "process %s finished: returncode=%d", proc, proc.returncode
-                    )
+                    if not proc._has_returncode.is_set():
+                        # If the process has not returned at this stage we need to hard
+                        # kill it to prevent it from hanging.
+                        logger.warning(
+                            "%s failed to exit cleanly.  Sending SIGKILL",
+                            proc,
+                            # The `any` call is to include a stacktrace if this
+                            # happened due to an exception but to omit it if this is
+                            # somehow happening outside of an exception context.
+                            exc_info=any(sys.exc_info()),
```

Comment:
`any` thing is clever (good clever).  I appreciate the comment.

Reply:
Can't take credit for it, though, as I simply borrowed it from you: https://github.com/ethereum/asyncio-run-in-process/blame/master/asyncio_run_in_process/main.py#L407 :-)